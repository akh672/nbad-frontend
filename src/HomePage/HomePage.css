@import url('https://fonts.googleapis.com/css2?family=Kanit:wght@100;200&family=Nova+Square&family=Outfit:wght@500;700&family=Roboto+Condensed:wght@400;625;900&display=swap');
.masterDashBoardContainer{
  font-family: Roboto Condensed;
}
.upperDashboardContainer {
  display: flex;
}

.verticalChartContainer {
  height: 55vh;
  width: 60%;
  padding: 20px;
  padding: 30px;
  margin: 10px 10px 10px 5px;
  position: relative;
}

.totaExpensesVsIncomeAmountperMonth {
  height: 58vh;
  width: 40%;
  max-width: 40%;
  padding: 20px;
  margin: 10px 5px 5px 5px;
  position: relative;
}

.downDashboardContainer {
  display: flex;
}

.pieChartContainer {
  position: relative;
  height: 50vh;
  padding: 30px;
  width: 60%;
  margin: 10px 10px 10px 5px;
}

.lineChartContainer {
  margin: 10px 5px 5px 5px;
  height: 50vh;
  width: 40%;
  position: relative;
  padding: 30px 0;
}

.incomExpanceDividerContainer {
  height: 100%;
  display: flex;
  justify-content: center;
  padding-left: 20%;
  flex-direction: column;
}

.fractionLabel {
  font-size: 40px;
  font-weight: 1000;
}

.fractionValue {
  color: #f1b92c;
}

.incomeFraction {
  margin-left: 40px;
}

.hrCont {
  width: 60%;
  height: 10px;
}

.hrCont > hr {
  height: 2px;
  background-color: black;
}

.headerForBoxes 
{
  position: absolute;
  top: 10px;
  font-size: 20px;
  font-weight: 600;
  left: 10px
  /* padding: 10px; */
}