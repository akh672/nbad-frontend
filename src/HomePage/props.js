import {
  getBudgetData,
  addBudgetData,
  createCatgory,
  deleteCategory,
  getAllExpensesAndDate,
} from '../Redux/action';

export const mapDispatchToProps = (dispatch) => ({
  getBudgetData: () => dispatch(getBudgetData()),
  addBudgetData: (budgetName, totalAmount, catagories) => dispatch(addBudgetData(budgetName, totalAmount, catagories)),
  createCatgory: (name,allocatedAmount,budgetId) => dispatch(createCatgory(name,allocatedAmount,budgetId)),
  deleteCategory: (_id) => dispatch(deleteCategory(_id)),
  getAllExpensesAndDate: () => dispatch(getAllExpensesAndDate()),
});

export const mapStateToProps = (state) => ({
  userRegistrationSuccessful: state.userRegistrationSuccessful,
  activeUserDetails: state.activeUserDetails,
  budgetsData: state.budgetsData,
  barChartExpnesesData: state.barChartExpnesesData,
  expensesData: state.expensesData,
});