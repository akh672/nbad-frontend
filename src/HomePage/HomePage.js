import React, { useEffect, useState } from 'react';
import {
  Chart as ChartJS,
  CategoryScale,
  LinearScale,
  BarElement,
  ArcElement,
  Title,
  Tooltip,
  Legend,
} from 'chart.js';
import Paper from '@mui/material/Paper';
import { Bar, Pie } from 'react-chartjs-2';
import InputLabel from '@mui/material/InputLabel';
import MenuItem from '@mui/material/MenuItem';
import Select from '@mui/material/Select';
import './HomePage.css';
import Navbar from '../Navbar';
// import faker from 'faker';
ChartJS.register(
  CategoryScale,
  LinearScale,
  BarElement,
  Title,
  Tooltip,
  Legend,
  ArcElement,
);








const HomePage = (props) => {

  const [budgetValue, setBudgetValue] = useState(0);
  const [totalExpenseValue, setExpenseValue] = useState(0);
  const [doubleBarLabels, setDoubleBarLabels] = useState([]);
  const [expensesForBarGraph, setExpensesForBarGraph] = useState([]);
  const [budgetForBarGraph, setBudgetForBarGraph] = useState([]);
  const [catagoriesNameArray, setCatalogNameArray] = useState([]);
  const [catagoryAmountArray, setCatalogAmountArray] = useState([]);
  const [amountPerCatagory, setAmountPerCatagory] = useState([])
  const {
    getBudgetData,
    budgetsData,
    getAllExpensesAndDate,
    expensesData,
    barChartExpnesesData,
  } = props;
  let labels = [];

  useEffect(() => {
    const fetchBudgetData = async () => {
      try {
        await getBudgetData();
        await getAllExpensesAndDate();
      }
      catch (er) {
        alert('Something went Wrong', er.message);
      }
    }
    fetchBudgetData();
  }, [])

  useEffect(() => {
    if (budgetsData && Object.keys(budgetsData).length > 0) {
      const {
        categories,
        totalAmount,
      } = budgetsData;
      // debugger
      let result = {};
      categories.forEach(category => {
        result[category.name] = result[category.name] ? result[category.name] + category.allocatedAmount : category.allocatedAmount;
      });
      console.log(result)
      setAmountPerCatagory(result);
      let categoriesValueCont = 0;
      let budgetAmount = [];
      let budgetNameArray = [];
      categories.map((budget) => {
        budgetAmount.push(budget.allocatedAmount);
        budgetNameArray.push(budget.name)
      });
      setCatalogAmountArray(budgetAmount);
      setCatalogNameArray(budgetNameArray);
      setBudgetValue(totalAmount);

    }
  }, [budgetsData]);

  useEffect(() => {
    let dateArray = [];
    let amountSpentArray = [];
if(budgetsData && Object.keys(budgetsData).length > 0){
  const {
    categories,
    totalAmount,
  } = budgetsData;
  const sortedData = barChartExpnesesData.sort((a, b) => new Date(a.date) - new Date(b.date));

  let decrementArrayBudget = [];
  let budgetLeft = totalAmount;
  let expensesDone = 0;
  sortedData.map((exp) => {
    dateArray.push(exp.date);
    decrementArrayBudget.push(budgetLeft - exp.totalAmountSpent)
    budgetLeft -= exp.totalAmountSpent;
    expensesDone += exp.totalAmountSpent;
    amountSpentArray.push(exp.totalAmountSpent);
  });
  console.log(amountSpentArray, decrementArrayBudget)
  setExpenseValue(expensesDone)
  setBudgetForBarGraph(decrementArrayBudget)
  setDoubleBarLabels(dateArray);
  setExpensesForBarGraph(amountSpentArray);
}

  }, [barChartExpnesesData]);


  return (<>
    <div className="masterDashBoardContainer">
      <Navbar title={'Dashboard'} />
      <div className="upperDashboardContainer">
        <Paper elevation={3} className="verticalChartContainer">
          <div className="headerForBoxes">EXPENSES and BUDGET LEFT Vs DATE</div>
          <Bar options={{
            maintainAspectRatio: false,
            responsive: true,
            plugins: {
              legend: {
                position: 'top',
              },
              // title: {
              //   display: true,
              //   text: 'Chart.js Bar Chart',
              // },
            },
            scales: {
              y: {
                beginAtZero: true,
                suggestedMax: Math.max([...expensesForBarGraph, ...budgetForBarGraph]) + 100,
              },
            },
          }} data={{
            labels: doubleBarLabels,
            datasets: [
              {
                label: 'Expenses',
                data: expensesForBarGraph,
                backgroundColor: '#fa6166',
              },
              {
                label: 'Budget Left',
                data: budgetForBarGraph,
                backgroundColor: '#f1b92c',
              },
            ],
          }} />
        </Paper>
        <Paper elevation={3} className='totaExpensesVsIncomeAmountperMonth'>
          <div className="headerForBoxes">TOTAL EXPENSES / TOTAL BUDGET</div>

          <div className="incomExpanceDividerContainer">
            <div className='fractionLabel'>
              Expenses: <span className='fractionValue'>{totalExpenseValue}</span>
            </div>
            <div className="hrCont">
              <hr />
            </div>
            <div className='fractionLabel'>
              Budget: <span className='fractionValue incomeFraction'>{budgetValue}</span>
            </div>
          </div>
        </Paper>
      </div>
      <div className="downDashboardContainer">
        <Paper elevation={3} className='pieChartContainer'>
          <div className="headerForBoxes">EXPENSES Per BUDGET CATAGORY</div>
          <Pie
            data={{
              labels: Object.keys(amountPerCatagory),
              datasets: [
                {
                  label: '# of Votes',
                  data: Object.values(amountPerCatagory),
                  backgroundColor: [
                    '#f1b92c',
                    '#fa6166',
                    '#4065cc',
                    'rgba(75, 192, 192, 0.2)',
                    'rgba(153, 102, 255, 0.2)',
                    'rgba(255, 159, 64, 0.2)',
                  ],
                  borderWidth: 1,
                },
              ],
            }}
            options={{
              responsive: true,
              maintainAspectRatio: false,
              tooltips: {
                enabled: true,
              },
              plugins: {
                datalabels: {
                  display: true,
                },
                htmlLegend: {
                  // ID of the container to put the legend in
                  containerID: 'legend-container',
                },
                legend: {
                  display: true,
                  position: 'right'
                }
              },
              hover: {
                animationDuration: 0,
              },
              animation: {
                animateScale: true,
                animateRotate: true,
                easing: 'easeInOutCirc',
              },
            }}
          />
        </Paper>
        <Paper elevation={3} className='lineChartContainer'>
          <div className="headerForBoxes">AMOUNT Vs CATAGORIES</div>
          <Bar options={{
            maintainAspectRatio: false,
            responsive: true,
            plugins: {
              legend: {
                position: 'top',
              },
            },
            scales: {
              y: {
                beginAtZero: true,
                suggestedMax: Math.max([...catagoryAmountArray]) + 100,
              },
            },
          }} data={{
            labels: catagoriesNameArray,
            datasets: [
              {
                label: 'Budget per Catagory',
                data: catagoryAmountArray,
                backgroundColor: '#4065cc',
              },
            ],
          }} />
        </Paper>
      </div>
    </div>
  </>)
}

export default HomePage;