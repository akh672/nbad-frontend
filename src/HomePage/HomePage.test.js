import React from 'react';
import { shallow, configure } from 'enzyme';
import Adapter from "@cfaester/enzyme-adapter-react-18";
configure({ adapter: new Adapter() });
import HomePage from './HomePage';

describe(`<HomePage/>`, () => {
  it('Should render <HomePage/> component without crashing', () => {
    const component = shallow(<HomePage/>);
    const tree = component.debug();
    expect(tree).toMatchSnapshot();
  });

  it(`Should render EXPENSES and BUDGET LEFT Vs DATE section`, () => {
    const component = shallow(<HomePage/>);
    expect(component.find('.verticalChartContainer').exists()).toBe(true);
  })
  
  it(`Should render TOTAL EXPENSES / TOTAL BUDGET section`, () => {
    const component = shallow(<HomePage/>);
    expect(component.find('.totaExpensesVsIncomeAmountperMonth').exists()).toBe(true);
  })
  
  it(`Should render EXPENSES Per BUDGET CATAGORY section`, () => {
    const component = shallow(<HomePage/>);
    expect(component.find('.pieChartContainer').exists()).toBe(true);
  })
  
  it(`Should render AMOUNT Vs CATAGORIES section`, () => {
    const component = shallow(<HomePage/>);
    expect(component.find('.lineChartContainer').exists()).toBe(true);
  })
});