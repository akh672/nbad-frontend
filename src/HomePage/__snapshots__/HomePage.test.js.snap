// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomePage/> Should render <HomePage/> component without crashing 1`] = `
"<Fragment>
  <div className=\\"masterDashBoardContainer\\">
    <Navbar title=\\"Dashboard\\" />
    <div className=\\"upperDashboardContainer\\">
      <ForwardRef(Paper) elevation={3} className=\\"verticalChartContainer\\">
        <div className=\\"headerForBoxes\\">
          EXPENSES and BUDGET LEFT Vs DATE
        </div>
        <ForwardRef options={{...}} data={{...}} />
      </ForwardRef(Paper)>
      <ForwardRef(Paper) elevation={3} className=\\"totaExpensesVsIncomeAmountperMonth\\">
        <div className=\\"headerForBoxes\\">
          TOTAL EXPENSES / TOTAL BUDGET
        </div>
        <div className=\\"incomExpanceDividerContainer\\">
          <div className=\\"fractionLabel\\">
            Expenses: 
            <span className=\\"fractionValue\\">
              0
            </span>
          </div>
          <div className=\\"hrCont\\">
            <hr />
          </div>
          <div className=\\"fractionLabel\\">
            Budget: 
            <span className=\\"fractionValue incomeFraction\\">
              0
            </span>
          </div>
        </div>
      </ForwardRef(Paper)>
    </div>
    <div className=\\"downDashboardContainer\\">
      <ForwardRef(Paper) elevation={3} className=\\"pieChartContainer\\">
        <div className=\\"headerForBoxes\\">
          EXPENSES Per BUDGET CATAGORY
        </div>
        <ForwardRef data={{...}} options={{...}} />
      </ForwardRef(Paper)>
      <ForwardRef(Paper) elevation={3} className=\\"lineChartContainer\\">
        <div className=\\"headerForBoxes\\">
          AMOUNT Vs CATAGORIES
        </div>
        <ForwardRef options={{...}} data={{...}} />
      </ForwardRef(Paper)>
    </div>
  </div>
</Fragment>"
`;
